MetaMask/metamask-extension
# Trigger token detection when user enables the setting

## **Description**

### Problem

Today when a user enables the token autodetection feature in settings, they still have to click "Refresh list" on the home page to actually trigger a detection.  We want the detection to automatically trigger when the setting is enabled.

### Solution

This feature almost worked already, but there was an issue with the order controllers were notified:
```
// preferences.js
this.store.updateState({ useTokenDetection: val });
this.tokenListController.start();
```

The `DetectTokensController` was subscribed to the feature being enabled.  But the `TokenListController` had not yet been started.  Therefore the token list used for detection was still empty.

I don't want to delay updating the setting until `TokenListController` is started.  So it feels most natural to fix this with event subscriptions.  `DetectTokensController` is now subscribed to changes in the `TokenListController` state, and will re-run a detection when the token list changes.  This also seems useful in case the token list changes for some other reason.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/1564

## **Manual testing steps**

1. Open a MM account that contains tokens on a non-mainnet chain.
2. Enable token autodetection in settings
3. Home page should say "N new token(s) found in this account"

## **Screenshots/Recordings**



### **Before**

https://github.com/MetaMask/metamask-extension/assets/3500406/b51c3a7d-3219-4400-98d0-fe8e5035d75b

### **After**

https://github.com/MetaMask/metamask-extension/assets/3500406/e2b0575d-06b0-4fac-bdd8-df82e2fad5e5

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
